libs/fairness/src/lib/util/calculateFairnessMetric.test.ts

```diff
@@ -0,0 +1,197 @@
+
+import { calculateFairnessMetric } from "./calculateFairnessMetric";
+import { FairnessModes } from "./FairnessMetrics";
+
+const closeTo = (expected: number, precision = 2) => ({
+  asymmetricMatch: (actual: number) =>
+    Math.abs(expected - actual) < Math.pow(10, -precision) / 2
+});
+describe("calculateFairnessMetric", () => {
+  it("should return NaN for no values", () => {
+    const mockValue = {
+      bins: [],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Difference);
+    const expectedResult = {
+      overall: Number.NaN
+    };
+    expect(result).toEqual(expectedResult);
+  });
+  it("should calculate difference without binBounds correctly", () => {
+    const mockValue = {
+      bins: [0.5, 0.2],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Difference);
+    const expectedResult = {
+      overall: 0.3
+    };
+    expect(result).toEqual(expectedResult);
+  });
+  it("should calculate difference (with overlap) correctly", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.25, upper: 0.55 },
+        { lower: 0.45, upper: 0.75 }
+      ],
+      bins: [0.3, 0.7],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Difference);
+    const expectedResult = {
+      bounds: { lower: 0, upper: closeTo(0.5) },
+      overall: closeTo(0.4)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should calculate difference (without overlap) correctly", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.25, upper: 0.35 },
+        { lower: 0.65, upper: 0.75 }
+      ],
+      bins: [0.3, 0.7],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Difference);
+    const expectedResult = {
+      bounds: { lower: closeTo(0.3), upper: closeTo(0.5) },
+      overall: closeTo(0.4)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should calculate difference with 3+ groups correctly", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.25, upper: 0.35 },
+        { lower: 0.65, upper: 0.75 },
+        { lower: 0.45, upper: 0.55 }
+      ],
+      bins: [0.3, 0.7, 0.5],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Difference);
+    const expectedResult = {
+      bounds: { lower: closeTo(0.1), upper: closeTo(0.5) },
+      overall: closeTo(0.4)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should calculate ratio without binBounds correctly", () => {
+    const mockValue = {
+      bins: [0.5, 0.2],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Ratio);
+    const expectedResult = {
+      overall: 0.4
+    };
+    expect(result).toEqual(expectedResult);
+  });
+  it("should calculate ratio (with overlap) correctly", () => {
```

Comment:
Again, please make sure that these work with the bin order reversed as well.

Reply:
Also, when it comes to overlap, there are two possible - full overlap (i.e. x1 < y1 and x2>y2) and partial (where the order of points is x1, y1, x2, y2). Those both need to be checked as well.